// sanity-plugin-markdown/lib/_chunks/plugin-ca10aa4c.js
var _templateObject;
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
import { useClient, PatchEvent, set, unset, defineType, definePlugin } from 'sanity';
import { jsx } from 'react/jsx-runtime';
import { lazy, useState, useEffect, useCallback, useMemo, Suspense } from 'react';
import styled from 'styled-components';
import { Box, Text } from '@sanity/ui';
const MarkdownInputStyles = styled(Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  & .CodeMirror.CodeMirror {\n    color: ", ";\n    border-color: ", ";\n    background-color: inherit;\n  }\n\n  & .cm-s-easymde .CodeMirror-cursor {\n    border-color: ", ";\n  }\n\n  & .editor-toolbar,\n  .editor-preview-side {\n    border-color: ", ";\n  }\n\n  & .CodeMirror-focused .CodeMirror-selected.CodeMirror-selected.CodeMirror-selected {\n    background-color: ", ";\n  }\n\n  & .CodeMirror-selected.CodeMirror-selected.CodeMirror-selected {\n    background-color: ", ";\n  }\n\n  & .editor-toolbar > * {\n    color: ", ";\n  }\n\n  & .editor-toolbar > .active,\n  .editor-toolbar > button:hover,\n  .editor-preview pre,\n  .cm-s-easymde .cm-comment {\n    background-color: ", ";\n  }\n\n  & .editor-preview {\n    background-color: ", ";\n\n    & h1,\n    h2,\n    h3,\n    h4,\n    h5,\n    h6 {\n      font-size: revert;\n    }\n\n    & ul,\n    li {\n      list-style: revert;\n      padding: revert;\n    }\n\n    & a {\n      text-decoration: revert;\n    }\n  }\n"])), _ref => {
  let {
    theme
  } = _ref;
  return theme.sanity.color.card.enabled.fg;
}, _ref2 => {
  let {
    theme
  } = _ref2;
  return theme.sanity.color.card.enabled.border;
}, _ref3 => {
  let {
    theme
  } = _ref3;
  return theme.sanity.color.card.enabled.fg;
}, _ref4 => {
  let {
    theme
  } = _ref4;
  return theme.sanity.color.card.enabled.border;
}, _ref5 => {
  let {
    theme
  } = _ref5;
  return theme.sanity.color.selectable.primary.hovered.bg;
}, _ref6 => {
  let {
    theme
  } = _ref6;
  return theme.sanity.color.card.enabled.bg;
}, _ref7 => {
  let {
    theme
  } = _ref7;
  return theme.sanity.color.card.enabled.fg;
}, _ref8 => {
  let {
    theme
  } = _ref8;
  return theme.sanity.color.card.enabled.bg;
}, _ref9 => {
  let {
    theme
  } = _ref9;
  return theme.sanity.color.card.enabled.bg;
});
const SimpleMdeReact = lazy(() => import('react-simplemde-editor'));
function useSimpleMdeReact() {
  const [mounted, setMounted] = useState(false);
  useEffect(() => {
    requestAnimationFrame(() => setMounted(true));
  }, []);
  return mounted ? SimpleMdeReact : null;
}
const defaultMdeTools = ["heading", "bold", "italic", "|", "quote", "unordered-list", "ordered-list", "|", "link", "image", "code", "|", "preview", "side-by-side"];
function MarkdownInput(props) {
  var _a;
  const {
    value = "",
    onChange,
    elementProps: {
      onBlur,
      onFocus,
      ref
    },
    reactMdeProps: {
      options: mdeCustomOptions,
      ...reactMdeProps
    } = {},
    schemaType
  } = props;
  const client = useClient({
    apiVersion: "2022-01-01"
  });
  const {
    imageUrl
  } = (_a = schemaType.options) != null ? _a : {};
  const imageUpload = useCallback((file, onSuccess, onError) => {
    client.assets.upload("image", file).then(doc => onSuccess(imageUrl ? imageUrl(doc) : "".concat(doc.url, "?w=450"))).catch(e => {
      console.error(e);
      onError(e.message);
    });
  }, [client, imageUrl]);
  const mdeOptions = useMemo(() => {
    return {
      autofocus: false,
      spellChecker: false,
      sideBySideFullscreen: false,
      uploadImage: true,
      imageUploadFunction: imageUpload,
      toolbar: defaultMdeTools,
      status: false,
      ...mdeCustomOptions
    };
  }, [imageUpload, mdeCustomOptions]);
  const handleChange = useCallback(newValue => {
    onChange(PatchEvent.from(newValue ? set(newValue) : unset()));
  }, [onChange]);
  const SimpleMdeReact = useSimpleMdeReact();
  return /* @__PURE__ */jsx(MarkdownInputStyles, {
    children: SimpleMdeReact && /* @__PURE__ */jsx(Suspense, {
      fallback: /* @__PURE__ */jsx(Box, {
        padding: 3,
        children: /* @__PURE__ */jsx(Text, {
          children: "Loading editor..."
        })
      }),
      children: /* @__PURE__ */jsx(SimpleMdeReact, {
        ...reactMdeProps,
        ref,
        value,
        onChange: handleChange,
        onBlur,
        onFocus,
        options: mdeOptions,
        spellCheck: false
      })
    })
  });
}
const markdownTypeName = "markdown";
const markdownSchemaType = defineType({
  type: "string",
  name: markdownTypeName,
  title: "Markdown",
  components: {
    input: MarkdownInput
  }
});
const markdownSchema = definePlugin(config => {
  return {
    name: "markdown-editor",
    schema: {
      types: [config && config.input ? {
        ...markdownSchemaType,
        components: {
          input: config.input
        }
      } : markdownSchemaType]
    }
  };
});
export { MarkdownInput, defaultMdeTools, markdownSchema, markdownSchemaType };
